internal/parse/parse.go

```diff
@@ -37,6 +37,70 @@ var (
 	errBadVolSrc    = errors.New("must set volume source")
 )
 
+// BuildOutputOptions contains parsed field generated by parsing build output string
+type BuildOutputOption struct {
+	Path     string
+	IsDir    bool
+	IsStdout bool
+}
+
+// GetBuildOutput is responsible for parsing custom build output argument.
+// Takes `buildOutput` as string and returns path where output has to be generated returns BuildOutputOptions
+func GetBuildOutput(buildOutput string) (BuildOutputOption, error) {
+	isDir := true
+	isStdout := false
+	typeSelected := false
+	pathSelected := false
+	path := ""
+	if len(buildOutput) == 1 && buildOutput == "-" {
+		// Feature parity with buildkit, output tar to stdout
+		// Read more here: https://docs.docker.com/engine/reference/commandline/build/#custom-build-outputs
+		return BuildOutputOption{"", false, true}, nil
+	}
+	if strings.Contains(buildOutput, ",") {
+		tokens := strings.SplitN(buildOutput, ",", 2)
+		if len(tokens) != 2 {
+			return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid build output options %q", buildOutput)
+		}
+		for _, option := range tokens {
+			arr := strings.SplitN(option, "=", 2)
+			if len(arr) != 2 {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid build output options %q", buildOutput)
+			}
+			if arr[0] == "type" && !typeSelected {
+				typeSelected = true
+				if arr[1] == "local" {
+					isDir = true
+				} else if arr[1] == "tar" {
+					isDir = false
+				} else {
+					return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid type %q selected for build output options %q", arr[1], buildOutput)
+				}
+			} else if arr[0] == "type" && pathSelected {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("Duplicate %q not supported", arr[0])
+			}
+			if arr[0] == "dest" && !pathSelected {
+				pathSelected = true
+				path = arr[1]
+			} else if arr[0] == "dest" && pathSelected {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("Duplicate %q not supported", arr[0])
+			}
+			if !typeSelected && !pathSelected {
```

Comment:
This works but it’s not quite obvious that it is the “unknown option name” requires maintaining this list of *Selected variables.

Also the two separate checks for `arr[0] == $value` seem unnecessary.

How about
```go
switch arr[0] {
case "type":
    if typeSelected { fail }
    typeSelected = true
    …
case "dest":
    if pathSelected { fail }
    pathSelected = true
    …
default:
    fail
}
```

or even
```go
// fieldsSeen := set<string> , I’m too lazy to use actual Go here. I guess a map[string]struct{}.
field := arr[0]
if field in fieldsSeen { fail }
fieldsSeen.Add(field)
switch field {
case "type":
    …
case "dest":
    …
default:
    fail
}
```

Reply:
fixed